lib/evaluate/fold.cc

```diff
@@ -156,35 +162,298 @@ ComplexPart FoldOperation(FoldingContext &context, ComplexPart &&complexPart) {
 
-template<typename T>
-Expr<T> FoldOperation(FoldingContext &context, FunctionRef<T> &&funcRef) {
-  ActualArguments args{std::move(funcRef.arguments())};
-  for (std::optional<ActualArgument> &arg : args) {
+namespace intrinsicHelper {
+// helpers to fold intrinsic function references
+// Define callable types used in a common utility that
+// takes care of array and cast/conversion aspects for elemental intrinsics
+
+template<typename TR, typename... TArgs>
+using ScalarFunc = std::function<Scalar<TR>(const Scalar<TArgs> &...)>;
+template<typename TR, typename... TArgs>
+using ScalarFuncWithContext =
+    std::function<Scalar<TR>(FoldingContext &, const Scalar<TArgs> &...)>;
+
+template<template<typename, typename...> typename WrapperType, typename TR,
+    typename... TA, std::size_t... I>
+static inline Expr<TR> FoldElementalIntrinsicHelper(FoldingContext &context,
+    FunctionRef<TR> &&funcRef, WrapperType<TR, TA...> func,
+    std::index_sequence<I...>) {
+  static_assert(
+      (... && IsSpecificIntrinsicType<TA>));  // TODO derived types for MERGE?
+  std::tuple<const std::optional<Scalar<TA>>...> scalars{
+      GetScalarConstantValue<TA>(*funcRef.arguments()[I]->value)...};
+  if ((... && std::get<I>(scalars).has_value())) {
+    if constexpr (std::is_same_v<WrapperType<TR, TA...>,
+                      ScalarFuncWithContext<TR, TA...>>) {
+      return Expr<TR>{Constant<TR>{func(context, *std::get<I>(scalars)...)}};
+    } else if constexpr (std::is_same_v<WrapperType<TR, TA...>,
+                             ScalarFunc<TR, TA...>>) {
+      return Expr<TR>{Constant<TR>{func(*std::get<I>(scalars)...)}};
+    }
+  }
+  // TODO: handle arrays when Constant<T> can represent them
+  return Expr<TR>{std::move(funcRef)};
+}
+
+template<typename TR, typename... TA>
+static Expr<TR> FoldElementalIntrinsic(FoldingContext &context,
+    FunctionRef<TR> &&funcRef, ScalarFunc<TR, TA...> func) {
+  return FoldElementalIntrinsicHelper<ScalarFunc, TR, TA...>(
+      context, std::move(funcRef), func, std::index_sequence_for<TA...>{});
+}
+template<typename TR, typename... TA>
+static Expr<TR> FoldElementalIntrinsic(FoldingContext &context,
+    FunctionRef<TR> &&funcRef, ScalarFuncWithContext<TR, TA...> func) {
+  return FoldElementalIntrinsicHelper<ScalarFuncWithContext, TR, TA...>(
+      context, std::move(funcRef), func, std::index_sequence_for<TA...>{});
+}
+}
+
+template<int KIND>
+Expr<Type<TypeCategory::Integer, KIND>> FoldOperation(FoldingContext &context,
+    FunctionRef<Type<TypeCategory::Integer, KIND>> &&funcRef) {
+  using namespace intrinsicHelper;
+  using T = Type<TypeCategory::Integer, KIND>;
+  for (std::optional<ActualArgument> &arg : funcRef.arguments()) {
     if (arg.has_value()) {
       *arg->value = FoldOperation(context, std::move(*arg->value));
     }
   }
   if (auto *intrinsic{std::get_if<SpecificIntrinsic>(&funcRef.proc().u)}) {
-    std::string name{intrinsic->name};
+    const std::string name{intrinsic->name};
```

Comment:
Why add the `const` qualifier to a local variable?

Reply:
I am afraid the code below will grow a lot and I want to give the guarantee to the reader when he sees "name" somewhere below that it was not changed in the code in-between. Plus, our style guideline says to use it generously.

Reply:
Yes, it does recommend generous use of `const`.  It makes sense to `const`-qualify pointers and references and class members as much as we can.  Const qualification on local variables is really rare, which is why I asked about it.

Reply:
I think it's reasonable to use it for mutable objects too (as opposed const on intrinsic types). You're providing a guarantee similar to const references that you won't call non-const methods on it.